### DISCUSSION TITLE
[Linux] Sleep until a specified real time

### DISCUSSION THREAD

**Discussion by 4e554c4c** [tokio-rs/tokio#4714] [Q&A] [answered]
See [this issue](https://github.com/mfreeborn/heliocron/issues/43) for all context.

Tokio provides [sleep_until](https://docs.rs/tokio/latest/tokio/time/fn.sleep_until.html) which sleeps until a given  _instant of a monotonic clock_. However, it does not provide any interfaces for sleeping until an _absolute instant of a realtime clock_ (using posix timer terminology).
This means that it is possible to wait for a certain amount of thread time to complete, but not wait until a certain wall time has passed. These can be drastically different, considering that many devices spend most of their time suspended—where wall time elapses but CPU time does not.

This is possible in unix using a combination of `CLOCK_REALTIME` and `TIMER_SETABSTIME`, see [this stackoverflow answer](https://stackoverflow.com/a/32152629/7903988) for more information. Is there any way to wait on an event like this using tokio? or without resorting to raw libc/librt calls?

**Comment by Darksonn** [comment, marked as answer]
You can do this using the `TimerFd` struct in the `tokio-timerfd` crate.

**Reply by 4e554c4c** [reply]
yep, after poking around a bit there doesn't seem to be a place for this in the tokio library, its timer wheel is not designed for wall clock timers.
I'd prefer a solution that didn't only work on linux, so I might make a posix compliant crate that does this.

**Reply by Darksonn** [reply]
Please let me know about it if you make such a crate.

**Reply by 4e554c4c** [reply]
https://git.sr.ht/~pounce/tokio-walltime
